### PR TITLE
Cylindrical gap treatment - M2D/M3D

### PR DESCRIPTION
**Pull request by gmalinve**

### CODE REVIEW

**Review by Samuelopez-ansys — COMMENTED** [review]
@gmalinve Why is not implemented inside Mesh.py? 

I think it is similar to "assign_rotational_layer" method. Then you will not need to duplicate the method for Maxwell 3D and 2D.

Please add an example in the docstring too

**Comment by Samuelopez-ansys** [comment]
Hi @gmalinve , 

@gmalinve Please check in the method if the design type is Maxwell, something like:

 ```
if self._app.design_type != "Maxwell 2D" and self._app.design_type != "Maxwell 3D":
            raise MethodNotSupportedError
```

File: pyaedt/modules/Mesh.py [hunk 1]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+        static_side_layers : int, optional
+            Number of mesh layers on the static Side.
+            The valid ranges are integers >= 1.
+            Option valid only for 3D design.
+            The default values is ``1``.
+
+        Returns
+        -------
+        :class:`pyaedt.modules.Mesh.MeshOperation` or bool
+            Mesh operation object or False if it fails.
```
Comment (MaxJPRey) [inline pyaedt/modules/Mesh.py:1385]:
```suggestion
            Mesh operation object or ``False`` if it fails.
```

File: pyaedt/modules/Mesh.py [hunk 2]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+            is automatically enabled.
+            For 3D design ``Clone Mesh`` option has to be enabled first.
+            The default is ``None``.
+        moving_side_layers : int, optional
+            Number of mesh layers on the moving side.
+            The valid ranges are integers >= 1.
+            Option valid only for 3D design.
+            The default values is ``1``.
+        static_side_layers : int, optional
+            Number of mesh layers on the static Side.
```
Comment (MaxJPRey) [inline pyaedt/modules/Mesh.py:1377]:
```suggestion
            Number of mesh layers on the static side.
```

File: pyaedt/modules/Mesh.py [hunk 3]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+                        "Name": meshop_name,
+                        "Objects": obj,
+                        "UseBandMappingAngle": use_band_mapping_angle,
+                        "BandMappingAngle": str(band_mapping_angle) + "deg",
+                    }
+                )
+            mop = MeshOperation(self, meshop_name, props, "CylindricalGap")
+            mop.create()
+            self.meshoperations.append(mop)
+            return mop
```
Comment (MaxJPRey) [inline pyaedt/modules/Mesh.py:1445]:
```suggestion
            mesh_operation = MeshOperation(self, meshop_name, props, "CylindricalGap")
            mesh_operation.create()
            self.meshoperations.append(mesh_operation)
            return mesh_operation
```

File: pyaedt/modules/Mesh.py [hunk 4]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+    def assign_cylindrical_gap(
+        self,
+        obj,
+        meshop_name=None,
+        band_mapping_angle=None,
+        clone_mesh=False,
+        moving_side_layers=1,
+        static_side_layers=1,
+    ):
+        """Assign a cylindrical gap for 2D/3D designs to enable use of Clone Mesh and associated Band Mapping Angle.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1348]:
```suggestion
        """Assign a cylindrical gap for a 2D or 3D design to enable a clone mesh and associated band mapping angle.
```

File: pyaedt/modules/Mesh.py [hunk 5]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+
+        Parameters
+        ----------
+        obj : int or str or :class:`pyaedt.modeler.object3d.Object3d`
+            Object to assign cylindrical gap to.
+        meshop_name : str, optional
+            Name of the mesh. The default is ``None``, in which
+            case the default name is used.
+        clone_mesh : bool, optional
+            Whether to clone the mesh.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1358]:
```suggestion
            Whether to clone the mesh. The default is ``False``.
```

File: _unittest/test_28_Maxwell3D.py [hunk 6]
```diff
@@ -793,3 +794,44 @@ def test_48_simplify_objects(self):
...
+            x.delete()
+            for x in m3d.mesh.meshoperations[:]
+            if x.type == "Cylindrical Gap Based" or x.type == "CylindricalGap"
+        ]
+        assert m3d.mesh.assign_cylindrical_gap(
+            "Band", meshop_name="cyl_gap_test", clone_mesh=True, band_mapping_angle=1
+        )
+        [
+            x.delete()
+            for x in m3d.mesh.meshoperations[:]
```
Comment (MaxJPRey) [inline _unittest/test_28_Maxwell3D.py:819]: Out of curiosity, why do we need to handle the slices here with `[:]` to delete some mesh operations?

File: pyaedt/modules/Mesh.py [hunk 7]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+            If ``True`` the solid bodies adjacent to the band are detected to identify the clone object.
+            The default is ``False``.
+            Option valid only for 3D design.
+        band_mapping_angle : int, optional
+            Band mapping angle in deg.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1363]:
```suggestion
            Band mapping angle in degrees.
```

File: pyaedt/modules/Mesh.py [hunk 8]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+        obj : int or str or :class:`pyaedt.modeler.object3d.Object3d`
+            Object to assign cylindrical gap to.
+        meshop_name : str, optional
+            Name of the mesh. The default is ``None``, in which
+            case the default name is used.
+        clone_mesh : bool, optional
+            Whether to clone the mesh.
+            If ``True`` the solid bodies adjacent to the band are detected to identify the clone object.
+            The default is ``False``.
+            Option valid only for 3D design.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1361]:
```suggestion
            Whether to clone the mesh. This parameter is valid only for 3D design.
            The default is ``False``. If ``True``, the solid bodies adjacent to the band
            are detected to identify the clone object.
```

File: pyaedt/modules/Mesh.py [hunk 9]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+        band_mapping_angle : int, optional
+            Band mapping angle in deg.
+            The recommended band mapping angle (the angle the rotor rotates in one time step) typically
+            equals the rotational speed multiplied by the time step.
+            Band mapping angle must be between 0.0005 and 3 deg.
+            For 2D design if a value is provided the option ``Use band mapping angle``
+            is automatically enabled.
+            For 3D design ``Clone Mesh`` option has to be enabled first.
+            The default is ``None``.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1370]:
```suggestion
            The band mapping angle must be between 0.0005 and 3 degrees.
            The default is ``None``.
            
            - For a 2D design, if a value is provided, the option ``Use band mapping angle``
              is automatically enabled.
            - For a 3D design, the ``Clone Mesh`` option has to be enabled first.
```

File: pyaedt/modules/Mesh.py [hunk 10]
```diff
@@ -1334,3 +1334,114 @@ def assign_density_control(self, names, refine_inside=True, maxelementlength=Non
...
+        moving_side_layers : int, optional
+            Number of mesh layers on the moving side.
+            The valid ranges are integers >= 1.
+            Option valid only for 3D design.
+            The default values is ``1``.
+        static_side_layers : int, optional
+            Number of mesh layers on the static side.
+            The valid ranges are integers >= 1.
+            Option valid only for 3D design.
+            The default values is ``1``.
```
Comment (PipKat) [inline pyaedt/modules/Mesh.py:1380]:
```suggestion
            The valid ranges are integers greater or equal to 1.
            This parameter is valid only for a 3D design.
            The default is ``1``.
        static_side_layers : int, optional
            Number of mesh layers on the static side.
            The valid ranges are integers greater than or equal to 1.
            This parameter is valid only for a 3D design.
            The default is ``1``.
```

**Pull request merged**